nativelib/src/main/scala/scala/scalanative/runtime/NativeExecutionContext.scala

```diff
@@ -0,0 +1,60 @@
+package scala.scalanative
+package runtime
+
+import scala.collection.mutable.ListBuffer
+import scala.concurrent.ExecutionContextExecutor
+import scala.scalanative.meta.LinktimeInfo.isMultithreadingEnabled
+
+object NativeExecutionContext {
+
+  /** Single-threaded queue based execution context. Each runable is executed
+   *  sequentially after termination of the main method
+   */
+  def queue: ExecutionContextExecutor = QueueExecutionContext
+
+  @deprecated(
+    "Use `queue` instead, since `global` may infer that it supports concurrent execution.",
+    since = "0.5.0"
+  )
+  def global: ExecutionContextExecutor = queue
+
+  object Implicits {
+    implicit val queue: ExecutionContextExecutor = NativeExecutionContext.queue
+  }
+
+  private object QueueExecutionContext extends ExecutionContextExecutor {
+    private val queue: ListBuffer[Runnable] = new ListBuffer
+    def execute(runnable: Runnable): Unit = queue += runnable
+    def reportFailure(t: Throwable): Unit = t.printStackTrace()
+
+    def hasNext: Boolean = queue.nonEmpty
+    def scheduled: Int = queue.size
+
+    def runNext(): Unit = if (hasNext) {
+      val runnable = queue.remove(0)
+      try runnable.run()
+      catch {
+        case t: Throwable =>
+          QueueExecutionContext.reportFailure(t)
+      }
+    }
+
+    def loop(): Unit = while (hasNext) runNext()
+  }
+
+  // The proxy methods here are defined to allow integrating projects
+  // to interact with our private execution context without leaking the abstraction
+
+  /** Checks if there is any scheduled runnable which can be executed */
+  private[runtime] def hasNext: Boolean = QueueExecutionContext.hasNext
+
+  /** Exuecute next available runnable in the queue. No-op if the queue is empty
+   */
+  private[runtime] def runNext(): Unit = QueueExecutionContext.runNext()
+
+  /** Returns the number of currently scheduled tasks */
+  private[runtime] def scheduled: Int = QueueExecutionContext.scheduled
+
+  /** Execute all the tasks in the queue until there is none left */
+  private[runtime] def loop(): Unit = QueueExecutionContext.loop()
```

Comment:
I'm glad we are considering these APIs, but maybe can this happen in a separate PR? I think this deserves a discussion. Cats Effect might also benefit from this, but we have also been successful without it—it is not a blocker.

In Scala Contributors[^1] we discussed adding some new APIs to Scala standard library that will better support event loops. For example a `Scheduler` API. This appears to have support from Scala Center, and Cats Effect team is also motivated for this change.

If those changes land in Scala 2.13 standard library, then it means single-thread Scala Native will be able to support tasks and timers out-of-the-box. So we should be careful to design this API so that we can make those changes without breaking backwards-compatibly.

[^1]: https://contributors.scala-lang.org/t/return-of-the-scala-actor/6041/8

Reply:
Oh, that's a good point. In such a case, I'll be putting this on hold, and I would use the old API instead in further integration of multithreading.

Reply:
I've reverted all the changes that expose new methods, I'll create a seperate PR with API extension